### ISSUE TITLE
Azure Functions Support

### ISSUE THREAD

**Issue by danielweil** [serverless-nextjs/serverless-next.js#1699]
Awesome work!

Do you guys plan to support other cloud providers as Azure?

**Comment by danielweil** [comment]
There is this git repository that uses this lib to convert it into azure Functions, but there is not support for getStaticProps && getInitialProps:

https://github.com/ShanonJackson/nextjs-azure-function

**Comment by dphang** [comment]
Yes it is on the roadmap as per some other issues, but I think first we need to refactor existing Lambda@Edge implementation. I am already simplifying the existing implementation so there is only one handler (no origin response handler) which uses S3 client to retrieve pages and perform static regeneration, fallbacks etc. 

However, we need to also decouple the AWS specific stuff from the actual routing code (e.g S3/SQS/CloudFront stuff is mixed into the handlers) and move it into a lambda-at-edge plugin. Currently there is `core` package which has some of this but would like to move more into it.

If you or someone is interested in the Azure part, it would be good to start with the following (as I am currently mainly the only one working on this, I am working on maintenance + Lambda/Lambda@Edge parts):

* Research how to deploy it, e.g Azure Blob Storage, Azure Functions, API Gateway equivalent whether it's through the GA version of Serverless Components, Terraform CDK, Serverless Framework. AFAICT for Serverless Framework I think they just handle the function deployments, but we would still need to upload. Ideally would try to stick to Serverless-standard stuff since that's most familiar
* Write the code in a separate `@sls-next/azure-plugin` package to bundle, package and deploy assuming all the assets are there (e.g similar to in `.serverless_nextjs`). For example you can start to create an Azure handler for pages/APIs (probably just a single one since Next.js has now optimized code splitting), image optimization that will be deployable. Though it will not be connected to any of the routing code yet.
* Then once we have the existing core/routing code decoupled, ideally it would just be a matter of inserting that code into the above handler.

**Comment by danielweil** [comment]
Hi Daniel, as I mentioned, in the comment above, I found out a repo that implements a wrapper to make it Azure compatible, but it doesn't work with server side features. It gets the output of api-lambdas folder and creates a router for the api part, passing it through the wrapper.

I am not sure I have the js skills to help out, or the time for it, but I will continue to have a look at it for now. Now I am trying to get a page with serverSideProps to run on Azure Function like it works with the api, but I think the router used does not work for this task, as it is from an old version. I will try to adapt the handler function to make this single page work, so I can make sure we can get through as I need in a project of mine.

If I am successfull, I can start having a look at a Azure replacement for S3 to host the static part.

What do you think?
Will keep you posted!
